File: sdk/python/examples/langchain/06_chat_history.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Chat History — multi-turn conversation with context.

Demonstrates:
    - Using create_agent which handles context natively
    - Tools for fact lookup
    - Running a single-turn query with the agent

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:6767/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain.agents import create_agent
from langchain_core.tools import tool
from langchain_openai import ChatOpenAI
from agentspan.agents import AgentRuntime


@tool
def recall_fact(topic: str) -> str:
    """Retrieve a stored fact about the given topic.

    Args:
        topic: The topic to look up (e.g., 'solar system', 'python').
    """
    facts = {
        "solar system": "The Solar System has 8 planets. Neptune is the farthest from the Sun.",
        "python": "Python was created by Guido van Rossum and first released in 1991.",
        "mars": "Mars is the fourth planet from the Sun and has two moons: Phobos and Deimos.",
        "earth": "Earth is the third planet from the Sun and the only known planet to harbor life.",
    }
    return facts.get(topic.lower(), f"No facts stored for '{topic}'.")


llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)
tools = [recall_fact]

graph = create_agent(
    llm,
    tools=tools,
    name="chat_history_agent",
    system_prompt="You are a helpful science assistant. Use tools to look up facts when needed.",
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        result = runtime.run(
            graph,
            "Which planet in the solar system is farthest from the Sun?",
        )
        print(f"Status: {result.status}")
        result.print_result()

        # Production pattern:
        # 1. Deploy once during CI/CD:
        # runtime.deploy(graph)
        # CLI alternative:
        # agentspan deploy --package examples.langchain.06_chat_history
        #
        # 2. In a separate long-lived worker process:
        # runtime.serve(graph)

File: sdk/typescript/examples/langchain/06-chat-history.ts
/**
 * Chat History -- maintaining conversation history across multiple turns.
 *
 * Demonstrates:
 *   - Accumulating HumanMessage/AIMessage pairs across turns
 *   - ChatOpenAI responding with full conversation context
 *   - Multi-turn conversation where the model recalls prior information
 *   - Running via AgentRuntime
 *
 * Requires: OPENAI_API_KEY environment variable
 */

import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, AIMessage, SystemMessage } from '@langchain/core/messages';
import type { BaseMessage } from '@langchain/core/messages';
import { RunnableLambda } from '@langchain/core/runnables';
import { AgentRuntime } from '../../src/index.js';

// ── Chat with history ────────────────────────────────────

class ChatWithHistory {
  private model: ChatOpenAI;
  private history: BaseMessage[];

  constructor() {
    this.model = new ChatOpenAI({ modelName: 'gpt-4o-mini', temperature: 0.3 });
    this.history = [
      new SystemMessage(
        'You are a friendly, helpful assistant. Remember everything the user tells you ' +
        'within this conversation and reference it when relevant. Keep responses concise.'
      ),
    ];
  }

  async chat(message: string): Promise<string> {
    this.history.push(new HumanMessage(message));

    const response = await this.model.invoke(this.history);
    const content = typeof response.content === 'string'
      ? response.content
      : JSON.stringify(response.content);

    this.history.push(new AIMessage(content));
    return content;
  }

  getHistoryLength(): number {
    return this.history.length;
  }
}

// ── Wrap as runnable for Agentspan ─────────────────────────

function createAgentRunnable() {
  const chat = new ChatWithHistory();

  const runnable = new RunnableLambda({
    func: async (input: { input: string }) => {
      const output = await chat.chat(input.input);
      return { output };
    },
  });

  // Add agentspan metadata for extraction
  (runnable as any)._agentspan = {
    model: 'openai/gpt-4o-mini',
    tools: [],
    framework: 'langchain',
  };

  return { runnable, chat };
}

async function main() {
  const turns = [
    'Hi! My name is Alex and I work in data science.',
    "I'm learning LangGraph for building AI agents.",
    "What's my name and what am I learning about?",
  ];

  // ── Run on agentspan ──────────────────────────────────────
  const runtime = new AgentRuntime();
  const { runnable, chat } = createAgentRunnable();

  try {
    for (let i = 0; i < turns.length; i++) {
      console.log(`\n--- Turn ${i + 1} ---`);
      console.log(`User: ${turns[i]}`);
      const result = await runtime.run(runnable, turns[i]);
      console.log(`Status: ${result.status}`);
      result.printResult();
      console.log(`(History: ${chat.getHistoryLength()} messages)`);
    }
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
